### ISSUE TITLE
Packaging/publication/naming (what to include in main stylesheet)

### ISSUE THREAD

**Issue by wrex** [Sepitus-exe/WKElementaryDark#59]
I like that we've broken out userscript styling and user overrides into separate files so the main stylesheet can be versioned independently (and will hopefully remain fairly stable).

We should start thinking about how to package and publish the separate stylesheets, though.

Just some rough ideas to kick off discussion:

- Publish `WKElementaryDark.css` as its own separately maintained and versioned stylesheet.

- Fold `WKED-wkof-menus.css` into the main stylesheet rather than as a separate file.

    Rationale: the Wanikani Open Framework is used by almost every user script in active use. Probably worth folding this into the main script (along with anything else _this_ prevalent).

- Keep `WKED-self-study.css`, `WKED-stroke-order.css`, and `WKED-ultimate-timeline.css` as separate stylesheets versioned and published independently. Users will need to pick and choose which of these (if any) that they want to install separately.

- Publish at least one non-default color scheme as an example for how to do your own theming. Propose we rename the "color-spins" folder "user-themes" and use a `.css` extension.

- We should probably start including the `@-moz-document` rules in the source files (both for the main stylesheet as well as userscript overrides and user-themes). I've just been manually adding "URLs in domain: wanikani.com" in the stylus editor, but we should probably an explicit `@-moz-document domain("wanikani.com") { ... }` (or whatever) around all the rules in revision control.

**Comment by Sepitus-exe** [comment]
> * Fold `WKED-wkof-menus.css` into the main stylesheet rather than as a separate file.

I was thinking of doing exactly this, it felt weird having to load 2 additional styles for a single script. I will add it as an issue.

> * Publish at least one non-default color scheme as an example for how to do your own theming. Propose we rename the "color-spins" folder "user-themes" and use a `.css` extension.

I agree with introducing some alternative color schemes, even a blank template might prove useful. But changing the folder name to "user-themes" seems a bit ambiguous. Theme, at least in my mind, evokes an image of an independently working stylesheet. Something like "user-palettes" might be more appropriate in our case.

> * We should probably start including the `@-moz-document` rules in the source files

It used to be a part of the file in the past. But it messed with indents on my side, so I decided to drop it.

It's not a direct solution, but I would prefer to leave the developer version as is and introduced a parallel file, which would reflect the state of the latest public release, including the `@-moz-document` wrapper and the `==UserStyle==` metadata.

---

We should also alter the userscripts to take their colors directly from the new palette files instead of the main theme file. It could be changed to take the colors in addition, rather than instead of, but working with `var(--USER, var(--ED, *color*))` could get messy really quickly.

**Comment by wrex** [comment]
> Something like "user-palettes" might be more appropriate in our case.

I'll name it "user-overrides" just in case we introduce any props for non-color things (spacing, etc.).

> It used to be a part of the file in the past. But it messed with indents on my side, so I decided to drop it.

Heh. I thought I accidentally deleted it while cutting-and-pasting between my editor and stylus. I need to see if I can figure out a more elegant solution than cut-n-paste.

Regardless, it's a minor nit. 

No work is the best work: I agree, let's just leave the dev versions without the `==UserStyle==` headers or `@moz-document` directives. I'll let you add those when publishing.

> We should also alter the userscripts to take their colors directly from the new palette files instead of the main theme file.

I'm not sure I follow, but I think we're both thinking the same thing: We're effectively publishing an API. 

We need to have the freedom to change _internal_ variables (e.g. palette colors) at will going forward. As long as we keep things internally consistent, those changes won't break anything in userscript (or user override) stylesheets.

But we are "publishing" an external interface with the "semantic" variables. We should be much more reticent to change, add to, or delete from those (even spelling or other naming changes). Changes to those props are effectively an API version change that will have ripple effects.

As long as userscript stylesheets only reference "semantic" props and don't access palette colors directly, I think maintenance is easy. On rare occasions, a userscript stylesheet may need to define its own colors for something we don't have a semantic variable for, but I think it's okay to leave that ad hoc.

I do, however, still think it's worthwhile to only have users override with `--USER-*` variables (not touching the `--ED-*` props). Having supported many things over the years, I like the bullet-proofing: I can pretty much guarantee that we'll field support requests from people not loading things in the correct order, or accidentally changing something they shouldn't have if we _don't_ do this.

> It could be changed to take the colors in addition, rather than instead of, but working with `var(--USER-*, var(--ED-*, color))` could get messy really quickly.

True. I'm playing with it now, though, and the pain is pretty localized and only done once. 

I'm starting to think this external/semantic vs. internal/don't-touch/"static" distinction is important. I'm beginning to think that we probably shouldn't use the same `--ED-*` naming convention for both, since the API is only the former.

**Comment by Sepitus-exe** [comment]
> I'm not sure I follow, but I think we're both thinking the same thing

> I do, however, still think it's worthwhile to only have users override with `--USER-*` variables

That's what I have been getting at. My intention was to remove the `--ED-*` overrides from all the userscript styles and replace them with `--USER-*` ones. The fallback colors should match the default theme, so there shouldn't be any reason to have the script inherit the same colors again. If there is an `--USER-*` variable, that is overriding some of the default `--ED-*` values, we can use it directly instead of going through the `--ED-*`. 

Additionally, if someone were to load only a user color palette and a script, everything should still work.

> > It could be changed to take the colors in addition, rather than instead of, but working with `var(--USER-*, var(--ED-*, color))` could get messy really quickly.

> True. I'm playing with it now, though, and the pain is pretty localized and only done once.

That might be true, but I don't see the benefit of still including the `--ED-*` inheritance. I would like to discourage the users from altering the main theme file, as their changes might get reverted with future updates that stylus automatically manages.

> I'm starting to think this external/semantic vs. internal/don't-touch/"static" distinction is important. I'm beginning to think that we probably shouldn't use the same `--ED-*` naming convention for both, since the API is only the former.

I wouldn't be opposed to changing the static prefix to something like `--Default` , `--Static`, `--Internal` or even something less telling but shorter like `--S`

---

We should also introduce some sort of variable that would work with extra-study img filters. Possibly even WK logo and other images. As things stand, the extra study img is an eyesore, when trying to move to different surface colors.

**Comment by wrex** [comment]
Ah!

I think we're accomplishing the same objective two different ways. Both of us want users (and userscript stylesheet writers) to leave these files alone and only create their own.

I was thinking WKElementaryDark.css itself as well as any userscript stylesheets should use any `--ED-*` props in their rules that they want. Those become the API.

We define those props at the top of the WKElementaryDark.css file like:

```css
:root {
  --ED-surface-1: var(--USER-surface-1, var(--EDI-gray-1));
  ...
}
```

Then users can override any of these as they wish in their own simple css files:

```css
:root {
  --ED-surface-1: #222222;
  ...
}
```

I've actually done all the work for this in the README branch already.

**Comment by wrex** [comment]
> We should also introduce some sort of variable that would work with extra-study img filters. Possibly even WK logo and other images. As things stand, the extra study img is an eyesore, when trying to move to different surface colors.

Agree. Take a look at how open-props does things. Please open an issue to track separately.

**Comment by wrex** [comment]
Also gonna nuke `--ED-brand-alt`. It was only used for "new" badges on the levels page, I think, and was just a slightly darker version of the brand color. I think just using the brand color suffices.

**Comment by Sepitus-exe** [comment]
I will take a look at all the changes tomorrow as Im short on time today. But it looks promising

**Comment by wrex** [comment]
All of this is included in Pull Request #65. Closing this issue so we can track it there.
